{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ee394075-2b86-49b4-ba67-42475549b0bf",
   "metadata": {},
   "source": [
    "# Visualize phylogenetic trees\n",
    "\n",
    "Import Python packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f41d2604-5caa-4451-92e5-b71271fdceff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "import itertools\n",
    "import math\n",
    "import os\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import ete3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c8e506-913d-442b-a355-a691c0850b78",
   "metadata": {},
   "source": [
    "Get variables from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9949903a-a253-442c-a4f5-853281ae8712",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tree_files = snakemake.input.trees\n",
    "state_files = snakemake.input.states\n",
    "all_csv = snakemake.input.all_csv\n",
    "alignment = snakemake.input.alignment\n",
    "deleted_csv = snakemake.input.deleted_csv\n",
    "comparator_map_csv = snakemake.input.comparator_map\n",
    "site_offset = snakemake.params.site_offset\n",
    "progenitors = snakemake.params.progenitors\n",
    "outgroups = snakemake.params.outgroups\n",
    "region_of_interest = snakemake.params.region_of_interest\n",
    "cat_colors = snakemake.params.cat_colors\n",
    "subcat_colors = snakemake.params.subcat_colors\n",
    "wuhan_hu_1_add_muts = snakemake.params.wuhan_hu_1_add_muts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24dbcbbf-d480-44d5-88ae-219d22a1d06d",
   "metadata": {},
   "source": [
    "Read data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e378dfd5-7bf6-4976-858e-d761a28e9ddd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_df = pd.read_csv(all_csv, na_filter=None)\n",
    "\n",
    "deleted_df = pd.read_csv(deleted_csv, na_filter=None)\n",
    "\n",
    "outgroup_map = pd.read_csv(comparator_map_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409e1979-accd-4de2-816d-92c30aa4851e",
   "metadata": {},
   "source": [
    "Read the reconstructed states and add the tip states, considering only sites where there are multiple identities.\n",
    "Now we make a matrix of substitutions of each node relative to each other.\n",
    "Note that in the general case this could get **really slow**, although for the small-ish alignment of similar sequences used here it's fine:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76776027-a9f5-4b58-94cc-a6810733eea4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "assert len(state_files) == len(tree_files) == len(progenitors)\n",
    "\n",
    "tip_to_seq = {s.id: list(str(s.seq).upper())\n",
    "              for s in Bio.SeqIO.parse(alignment, 'fasta')}\n",
    "\n",
    "tip_states = (\n",
    "    pd.DataFrame.from_dict(tip_to_seq, orient='index')\n",
    "    .rename_axis('Node')\n",
    "    .reset_index()\n",
    "    .melt(id_vars='Node',\n",
    "          var_name='Site',\n",
    "          value_name='State',\n",
    "          )\n",
    "    .assign(Site=lambda x: x['Site'] + 1)\n",
    "    )\n",
    "\n",
    "subs_matrices = {}\n",
    "\n",
    "for progenitor, state_file in zip(progenitors, state_files):\n",
    "    internal_states = (\n",
    "        pd.read_csv(state_file,\n",
    "                    sep='\\t',\n",
    "                    comment='#',\n",
    "                    usecols=['Node', 'Site', 'State'])\n",
    "        )\n",
    "\n",
    "    states = (\n",
    "        internal_states\n",
    "        .append(tip_states)\n",
    "        .assign(Site=lambda x: x['Site'] + site_offset,\n",
    "                n_states_at_site=lambda x: x.groupby('Site')['State'].transform('nunique'),\n",
    "                )\n",
    "        .query('n_states_at_site > 1')\n",
    "        .drop(columns='n_states_at_site')\n",
    "        )\n",
    "\n",
    "    states_dict = states.set_index(['Node', 'Site'])['State'].to_dict()\n",
    "\n",
    "    nodes = states['Node'].unique().tolist()\n",
    "    sites = sorted(states['Site'].unique())\n",
    "\n",
    "    subs_matrix = {}  # keyed by (parent, descendant)\n",
    "    for n1, n2 in itertools.product(nodes, nodes):\n",
    "        subs = []\n",
    "        for site in sites:\n",
    "            nt1 = states_dict[(n1, site)]\n",
    "            nt2 = states_dict[(n2, site)]\n",
    "            if nt1 != nt2:\n",
    "                if nt1 in {'A', 'C', 'G', 'T'} and nt2 in {'A', 'C', 'G', 'T'}:\n",
    "                    subs.append(f\"{nt1}{site}{nt2}\")\n",
    "        subs_matrix[(n1, n2)] = ', '.join(subs)\n",
    "    subs_matrices[progenitor] = subs_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3871a563-1c6d-4cab-b790-997618387059",
   "metadata": {},
   "source": [
    "Get annotations of which mutations are to an outgroup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "580191ca-cd7e-4ef9-8ffe-86c9904c0dee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "subs_to_outgroup = {outgroup: set(outgroup_map\n",
    "                                  .assign(sub=lambda x: x['site'].astype(str) + x[outgroup])\n",
    "                                  ['sub']\n",
    "                                  )\n",
    "                    for outgroup in outgroups}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de13b90e-f260-417d-82ac-0947e0ecef93",
   "metadata": {},
   "source": [
    "List information on the possible progenitors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a360f121-f971-4595-9f39-145750d91c2f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', 1000)\n",
    "\n",
    "progenitor_info = (\n",
    "    all_df\n",
    "    .query('representative_strain in @progenitors')\n",
    "    .assign(all_strains_dates=lambda x: x['all_strains_dates'].str.split(', '))\n",
    "    [['substitutions', 'representative_strain', 'nstrains', 'all_strains_dates']]\n",
    "    .set_index(['substitutions', 'representative_strain', 'nstrains'])\n",
    "    .explode('all_strains_dates')\n",
    "    .assign(strain=lambda x: x['all_strains_dates'].str.split(' \\(').str[0],\n",
    "            date=lambda x: x['all_strains_dates'].str.split(' \\(').str[1].str[: -1],\n",
    "            )\n",
    "    .drop(columns='all_strains_dates')\n",
    "    )\n",
    "\n",
    "display(progenitor_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea359c0-bb36-4a53-b786-abf1b584a268",
   "metadata": {},
   "source": [
    "Set up labels on possible progenitors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f1fb14-2d3d-4371-85eb-8e9e73aa16d1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "progenitor_dates = progenitor_info.set_index('strain')['date'].to_dict()\n",
    "progenitor_subs = progenitor_info.reset_index().set_index('strain')['substitutions'].to_dict()\n",
    "\n",
    "wu_hu_1_subs = {int(s[1: -1]): (s[0], s[-1]) for s in wuhan_hu_1_add_muts}\n",
    "\n",
    "node_labels = {}\n",
    "for progenitor in progenitors:\n",
    "    subs = [s for s in progenitor_subs[progenitor].split(',') if s]\n",
    "    \n",
    "    subs_wu_1 = []\n",
    "    sites_added = set()\n",
    "    for s in subs:\n",
    "        site, wt, mut = int(s[1: -1]), s[0], s[-1]\n",
    "        if site not in wu_hu_1_subs:\n",
    "            subs_wu_1.append((site, wt, mut))\n",
    "        else:\n",
    "            mut = wu_hu_1_subs[site][-1]\n",
    "            if mut != wt:\n",
    "                subs_wu_1.append((site, wt, mut))\n",
    "        sites_added.add(site)\n",
    "    for site, (wt, mut) in wu_hu_1_subs.items():\n",
    "        if site not in sites_added:\n",
    "            subs_wu_1.append((site, wt, mut))\n",
    "    subs_wu_1 = [f\"{wt}{site}{mut}\" for site, wt, mut in sorted(subs_wu_1)]\n",
    "        \n",
    "    node_labels[progenitor] = [\n",
    "            progenitor.replace('hCoV-19/', '') + f\" ({progenitor_dates[progenitor]})\",\n",
    "            'mutations from proCoV2 (Kumar et al): ' + (', '.join(subs) if subs else 'none'),\n",
    "            'mutations from Wuhan-Hu-1: ' + (', '.join(subs_wu_1) if subs else 'none'),\n",
    "            ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06cfb783-77cb-4650-a10f-c8ccdff5b723",
   "metadata": {},
   "source": [
    "This next line enables Jupyter notebook rendering of trees in `ete3`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e53af0-0e25-460a-a988-6d8e3b55d58e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to enable jupyter notebook rendering: \n",
    "os.environ['QT_QPA_PLATFORM'] = 'offscreen'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b87da384-5573-487b-9082-99c52ce11eaa",
   "metadata": {},
   "source": [
    "Add in deleted sequences at all \"plausible\" locations, adding proportional to number of nodes at each location at which it's consistent, and making the category labels the sub categories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "521552ac-cca7-4c5a-9753-8cc4528560cf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "start, end = region_of_interest['start'], region_of_interest['end']\n",
    "\n",
    "def is_compatible(all_subs, subs):\n",
    "    all_subs = {s for s in all_subs.split(',')\n",
    "                if s and start <= int(s[1 : -1]) <= end}\n",
    "    return subs == all_subs\n",
    "\n",
    "all_plus_deleted_df = (\n",
    "    all_df\n",
    "    .assign(deleted=0)\n",
    "    .rename(columns={c: c.split('_')[1] if c.startswith('subcat_') else c\n",
    "                     for c in all_df.columns})\n",
    "    )\n",
    "for tup in deleted_df.itertuples():\n",
    "    subs_str, n = tup.substitutions, tup.nstrains\n",
    "    subs = {s for s in subs_str.split(',') if s}\n",
    "    all_plus_deleted_df = (\n",
    "        all_plus_deleted_df\n",
    "        .assign(compatible=lambda x: x['substitutions'].apply(is_compatible, args=(subs,)))\n",
    "        )\n",
    "    compatible_df = all_plus_deleted_df[['substitutions', 'nstrains', 'compatible']].query('compatible')\n",
    "    n_tot = compatible_df['nstrains'].sum()\n",
    "    print(f\"There are {len(compatible_df)} sequences comprising {n_tot} strains compatible \"\n",
    "          f\"with the {n} deleted sequences with the following substitutions: {subs_str}\")\n",
    "    if n_tot == 0:\n",
    "        continue\n",
    "    all_plus_deleted_df = (\n",
    "        all_plus_deleted_df\n",
    "        .assign(deleted=lambda x: x['deleted'] + x['nstrains'] * n / n_tot * x['compatible'].astype(int))\n",
    "        )\n",
    "all_plus_deleted_df = (\n",
    "    all_plus_deleted_df\n",
    "    .assign(nstrains=lambda x: x['nstrains'] + x['deleted'])\n",
    "    .rename(columns={'deleted': 'deleted early Wuhan'})\n",
    "    .drop(columns='compatible')\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d836f4df-3014-4eb6-9477-c5b7ae2b977d",
   "metadata": {},
   "source": [
    "Modify `all_df` to just have categories as columns, and also to split the `Wuhan` category into `Huanan Seafood Market` and `other Wuhan`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1bed3f6-eb3c-4509-9855-8f3d04782f28",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "convert_cats_to_subcats = {'Wuhan': ['Huanan Seafood Market', 'other Wuhan']}\n",
    "\n",
    "for cat, subcats in convert_cats_to_subcats.items():\n",
    "    assert all(all_df[f\"cat_{cat}\"] == all_df[[f\"subcat_{subcat}\" for subcat in subcats]].sum(axis=1))\n",
    "    all_df_cat_cols = (\n",
    "        all_df\n",
    "        .drop(columns=f\"cat_{cat}\")\n",
    "        .rename(columns={f\"subcat_{subcat}\": subcat.replace('subcat_', 'cat_')\n",
    "                         for subcat in subcats}\n",
    "                )\n",
    "        )\n",
    "    new_cat_colors = {}\n",
    "    for key, val in cat_colors.items():\n",
    "        if key == cat:\n",
    "            for subcat in subcats:\n",
    "                new_cat_colors[subcat] = subcat_colors[subcat]\n",
    "        else:\n",
    "            new_cat_colors[key] = val\n",
    "    cat_colors = new_cat_colors\n",
    "\n",
    "all_df_cat_cols = all_df_cat_cols.rename(columns={c: c.split('_')[1] if c.startswith('cat_') else c\n",
    "                                                  for c in all_df.columns})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a53105-65a7-46fd-9d21-bfaeee76f328",
   "metadata": {},
   "source": [
    "Draw the trees:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "471a0505-cd4c-4762-88a4-601ea5a102bd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_PieChartFace(node_annotation, nodesizescale, locations):\n",
    "    nstrains = node_annotation['nstrains']\n",
    "    loc_sum = sum(node_annotation[loc] for loc in locations)\n",
    "    assert numpy.allclose(nstrains, loc_sum, atol=1e-3), f\"{nstrains=}\\n{loc_sum=}\\n{locations=}\\n{node_annotation=}\"\n",
    "    radius = nodesizescale * math.sqrt(node_annotation['nstrains'])\n",
    "    percents = []\n",
    "    colors = []\n",
    "    for loc, color in locations.items():\n",
    "        percents.append(100 * node_annotation[loc] / nstrains)\n",
    "        colors.append(color)\n",
    "    return ete3.PieChartFace(percents=percents,\n",
    "                             colors=colors,\n",
    "                             width=radius,\n",
    "                             height=radius,\n",
    "                             #line_color='black',\n",
    "                             )\n",
    "\n",
    "def get_pretty_tree(treefile,\n",
    "                    df,\n",
    "                    subs_matrix,\n",
    "                    progenitor,\n",
    "                    to_outgroup_muts,\n",
    "                    locations,\n",
    "                    widthscale=1350,\n",
    "                    heightscale=1,\n",
    "                    nodesizescale=35,\n",
    "                    label_nodes=False,\n",
    "                    label_fontsize=25,\n",
    "                    branch_linewidth=4,\n",
    "                    mut_label_color='black',\n",
    "                    mut_to_outgroup_label_color='#CC79A7',\n",
    "                    collapse_below={'C28144T': ['colapsed clade B', 'Wuhan-Hu-1']},\n",
    "                    node_label_color='#999999',\n",
    "                    node_labels={},\n",
    "                    title_fontsize=35,\n",
    "                    draw_legend=True,\n",
    "                    ):\n",
    "    \"\"\"Returns `(tree, tree_style)`.\"\"\"\n",
    "    annotations = df.set_index('representative_strain').to_dict(orient='index')\n",
    "    for loc in locations:\n",
    "        if loc not in list(annotations.values())[0]:\n",
    "            raise ValueError(f\"annotations missing {loc}:\\n{list(annotations.values())[0]}\")\n",
    "   \n",
    "    t = ete3.Tree(treefile, format=1)\n",
    "\n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = False  # add tip names manually\n",
    "     \n",
    "    nstyle_dict = {'hz_line_width': branch_linewidth,\n",
    "                   'vt_line_width': branch_linewidth,\n",
    "                   'hz_line_color': 'black',\n",
    "                   'vt_line_color': 'black',\n",
    "                   'size': 0}\n",
    "\n",
    "    # label nodes\n",
    "    for n in t.traverse():\n",
    "        if n != t:\n",
    "            subs = subs_matrix[(n.up.name, n.name)]\n",
    "            icol = 0\n",
    "            for sub in subs.split(', '):\n",
    "                if sub:\n",
    "                    to_outgroup = sub[1:] in to_outgroup_muts\n",
    "                    n.add_face(ete3.TextFace(f\"{sub}  \",\n",
    "                                             fsize=label_fontsize,\n",
    "                                             fgcolor=(mut_to_outgroup_label_color\n",
    "                                                      if to_outgroup else\n",
    "                                                      mut_label_color),\n",
    "                                             bold=to_outgroup,\n",
    "                                             ),\n",
    "                               column=icol,\n",
    "                               position='branch-top',\n",
    "                               )\n",
    "                    icol += 1\n",
    "        nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "        if n.is_leaf() or n.name == progenitor:\n",
    "            n.add_face(get_PieChartFace(annotations[n.name], nodesizescale, locations),\n",
    "                       column=0,\n",
    "                       position='branch-right',\n",
    "                       )\n",
    "            if label_nodes:\n",
    "                n.add_face(ete3.TextFace(annotations[n.name]['substitutions'],\n",
    "                                         tight_text=True,\n",
    "                                         fsize=labelfontsize,\n",
    "                                         ),\n",
    "                           column=0,\n",
    "                           position='branch-right',\n",
    "                           )\n",
    "        n.set_style(nstyle)\n",
    "        \n",
    "    # set dummy node for root, this allows us to put\n",
    "    # progenitor at base of root branch\n",
    "    progenitor_n = t.search_nodes(name=progenitor)\n",
    "    assert len(progenitor_n) == 1\n",
    "    progenitor_n = progenitor_n[0]\n",
    "    dummy_outgroup = progenitor_n.add_child(name='dummy',\n",
    "                                            dist=1e-5)\n",
    "    t.set_outgroup(dummy_outgroup)\n",
    "    assert len(progenitor_n.children) == 1\n",
    "    progenitor_child = progenitor_n.children[0]\n",
    "    nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "    progenitor_child.set_style(nstyle)\n",
    "    if progenitor_n.faces:\n",
    "        sub = subs_matrix[(progenitor_n.up.name, progenitor)]\n",
    "        if sub:\n",
    "            assert len(sub.split(',')) == 1\n",
    "            sub_flipped = f\"{sub[-1]}{sub[1: -1]}{sub[0]}\"\n",
    "            to_outgroup = sub_flipped[1:] in to_outgroup_muts\n",
    "            progenitor_child.add_face(ete3.TextFace(f\"{sub_flipped}  \",\n",
    "                                                    fsize=label_fontsize,\n",
    "                                                    fgcolor=(mut_to_outgroup_label_color\n",
    "                                                             if to_outgroup else\n",
    "                                                             mut_label_color),\n",
    "                                                    bold=to_outgroup,\n",
    "                                                    ),\n",
    "                                      column=0,\n",
    "                                      position='branch-top',\n",
    "                                      )\n",
    "        delattr(progenitor_n, '_faces')\n",
    "    empty_nstyle = ete3.NodeStyle(hz_line_color='white',\n",
    "                                  vt_line_color='white',\n",
    "                                  size=0)\n",
    "    dummy_outgroup.set_style(empty_nstyle)\n",
    "    dummy_outgroup.up.set_style(empty_nstyle)\n",
    "    progenitor_n.add_face(get_PieChartFace(annotations[progenitor], nodesizescale, locations),\n",
    "                          column=0,\n",
    "                          position='branch-right',\n",
    "                          )\n",
    "\n",
    "    t.ladderize()\n",
    "    \n",
    "    # collapse tree at indicated nodes\n",
    "    for below_subs, txt in collapse_below.items():\n",
    "        collapse_n = [n for n in t.traverse() if n.name and n.up and n.up.name and\n",
    "                      n.name != 'dummy' and subs_matrix[(n.up.name, n.name)] == below_subs]\n",
    "        if len(collapse_n) > 1:\n",
    "            raise ValueError(f\"more than one node to collapse for {below_subs}\")\n",
    "        elif len(collapse_n) == 0:\n",
    "            raise ValueError(f\"no nodes to collapse for {below_subs}\")\n",
    "        collapse_n = collapse_n[0]\n",
    "        collapse_nstrains = 0\n",
    "        collapse_annotations = {key: 0 for key in ['nstrains', *locations]}\n",
    "        for n in [collapse_n] + collapse_n.get_descendants():\n",
    "            if n.is_leaf():\n",
    "                collapse_annotations['nstrains'] += annotations[n.name]['nstrains']\n",
    "                for location in locations:\n",
    "                    collapse_annotations[location] += annotations[n.name][location]\n",
    "            if n != collapse_n:\n",
    "                n.detach()\n",
    "        nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "        collapse_n.set_style(nstyle)\n",
    "        collapse_n.add_face(get_PieChartFace(collapse_annotations, nodesizescale, locations),\n",
    "                            column=0,\n",
    "                            position='branch-right',\n",
    "                            )\n",
    "        for face_txt in txt:\n",
    "            collapse_n.add_face(ete3.TextFace(f\" {face_txt}\",\n",
    "                                              fsize=label_fontsize,\n",
    "                                              fgcolor=node_label_color,\n",
    "                                              bold=True,\n",
    "                                              ),\n",
    "                                column=1,\n",
    "                                position='branch-right',\n",
    "                                )\n",
    "    \n",
    "    # label specified non-progenitor nodes\n",
    "    for name, labels in node_labels.items():\n",
    "        if name == progenitor:\n",
    "            continue\n",
    "        n = t.search_nodes(name=name)\n",
    "        assert len(n) == 1\n",
    "        n = n[0]\n",
    "        for face_txt in labels[: 1]:\n",
    "            n.add_face(ete3.TextFace(f\" {face_txt}\",\n",
    "                                     fsize=label_fontsize,\n",
    "                                     fgcolor=node_label_color,\n",
    "                                     bold=True,\n",
    "                                     ),\n",
    "                       column=1,\n",
    "                       position='branch-right',\n",
    "                       )\n",
    "            \n",
    "    # add title\n",
    "    if progenitor in node_labels:\n",
    "        title_labels = node_labels[progenitor]\n",
    "    else:\n",
    "        title_labels = [progenitor]\n",
    "    for i, face_txt in enumerate(title_labels):\n",
    "        if i == 0:\n",
    "            prefix = 'progenitor as '\n",
    "        else:\n",
    "            prefix = ''\n",
    "        ts.title.add_face(ete3.TextFace(prefix + face_txt,\n",
    "                                        fsize=title_fontsize,\n",
    "                                        fgcolor='black'),\n",
    "                          column=0,\n",
    "                          )\n",
    "    ts.title.add_face(ete3.TextFace('', fsize=title_fontsize),\n",
    "                      column=0)\n",
    "    \n",
    "    ts.show_scale = False\n",
    "    height = t.get_farthest_node()[1]\n",
    "    ts.scale = widthscale / height\n",
    "    ts.branch_vertical_margin = heightscale\n",
    "    \n",
    "    if draw_legend:\n",
    "        for loc, color in locations.items():\n",
    "            # add padding\n",
    "            for col in [0, 1]:\n",
    "                ts.legend.add_face(ete3.RectFace(0.5 * nodesizescale,\n",
    "                                                 0.5 * nodesizescale,\n",
    "                                                 'white', 'white'),\n",
    "                                   column=col)\n",
    "            # add legend\n",
    "            ts.legend.add_face(ete3.CircleFace(1.2 * nodesizescale,\n",
    "                                               color),\n",
    "                               column=0)\n",
    "            ts.legend.add_face(ete3.TextFace(' ' + loc,\n",
    "                                             fsize=title_fontsize),\n",
    "                               column=1)\n",
    "        ts.legend_position = 3  # 3 is bottom left\n",
    "    \n",
    "    return t, ts\n",
    "\n",
    "assert len(progenitors) == len(tree_files)\n",
    "dirname = 'results/phylogenetics/tree_images'\n",
    "os.makedirs(dirname, exist_ok=True)\n",
    "for outgroup, (progenitor, tree_file) in itertools.product(outgroups,\n",
    "                                                           zip(progenitors, tree_files)\n",
    "                                                           ):\n",
    "    progenitor_str = progenitor.replace('/', '%')\n",
    "        \n",
    "    t, ts = get_pretty_tree(tree_file,\n",
    "                            df=all_df_cat_cols,\n",
    "                            subs_matrix=subs_matrices[progenitor],\n",
    "                            progenitor=progenitor,\n",
    "                            to_outgroup_muts=subs_to_outgroup[outgroup],\n",
    "                            node_labels=node_labels,\n",
    "                            locations=cat_colors,\n",
    "                            )\n",
    "    display(t.render('%%inline', tree_style=ts, w=300))\n",
    "    treefile = os.path.join(dirname, f\"{progenitor_str}_{outgroup}_without_deleted_seqs.pdf\")\n",
    "    print(f\"Saving to {treefile}\\n\")\n",
    "    t.render(treefile, tree_style=ts)\n",
    "    \n",
    "    t_deleted, ts_deleted = get_pretty_tree(\n",
    "                            tree_file,\n",
    "                            df=all_plus_deleted_df,\n",
    "                            subs_matrix=subs_matrices[progenitor],\n",
    "                            progenitor=progenitor,\n",
    "                            to_outgroup_muts=subs_to_outgroup[outgroup],\n",
    "                            node_labels=node_labels,\n",
    "                            locations=subcat_colors,\n",
    "                            )\n",
    "    display(t_deleted.render('%%inline', tree_style=ts_deleted, w=300))\n",
    "    treefile = os.path.join(dirname, f\"{progenitor_str}_{outgroup}_with_deleted_seqs.pdf\")\n",
    "    print(f\"Saving to {treefile}\\n\")\n",
    "    t_deleted.render(treefile, tree_style=ts_deleted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "040049df-28d0-485e-8481-253cca700016",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
