{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ee394075-2b86-49b4-ba67-42475549b0bf",
   "metadata": {},
   "source": [
    "# Visualize phylogenetic trees\n",
    "\n",
    "Import Python packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f41d2604-5caa-4451-92e5-b71271fdceff",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:06:41.329310Z",
     "iopub.status.busy": "2021-06-11T21:06:41.328995Z",
     "iopub.status.idle": "2021-06-11T21:06:41.333473Z",
     "shell.execute_reply": "2021-06-11T21:06:41.332696Z",
     "shell.execute_reply.started": "2021-06-11T21:06:41.329272Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "import itertools\n",
    "import math\n",
    "import os\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import ete3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c8e506-913d-442b-a355-a691c0850b78",
   "metadata": {},
   "source": [
    "Get variables from `snakemake`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "256d1901-8eb3-4515-a492-eb8cfe830f72",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:06:41.335257Z",
     "iopub.status.busy": "2021-06-11T21:06:41.335072Z",
     "iopub.status.idle": "2021-06-11T21:06:41.341626Z",
     "shell.execute_reply": "2021-06-11T21:06:41.340991Z",
     "shell.execute_reply.started": "2021-06-11T21:06:41.335235Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "######## snakemake preamble start (automatically inserted, do not edit) ########\n",
    "import sys; sys.path.extend(['/fh/fast/bloom_j/software/miniconda3/envs/SARS-CoV-2_PRJNA612766/lib/python3.8/site-packages', '/home/jbloom/SARS-CoV-2_PRJNA612766/notebooks']); import pickle; snakemake = pickle.loads(b\"\\x80\\x04\\x95\\x8e,\\x00\\x00\\x00\\x00\\x00\\x00\\x8c\\x10snakemake.script\\x94\\x8c\\tSnakemake\\x94\\x93\\x94)\\x81\\x94}\\x94(\\x8c\\x05input\\x94\\x8c\\x0csnakemake.io\\x94\\x8c\\nInputFiles\\x94\\x93\\x94)\\x81\\x94(\\x8c7results/phylogenetics/all_hCoV-19%USA%WA1%2020.treefile\\x94\\x8cDresults/phylogenetics/all_hCoV-19%Guangdong%HKU-SZ-002%2020.treefile\\x94\\x8c@results/phylogenetics/all_hCoV-19%Shandong%LY005-2%2020.treefile\\x94\\x8c4results/phylogenetics/all_hCoV-19%USA%WA1%2020.state\\x94\\x8cAresults/phylogenetics/all_hCoV-19%Guangdong%HKU-SZ-002%2020.state\\x94\\x8c=results/phylogenetics/all_hCoV-19%Shandong%LY005-2%2020.state\\x94\\x8c&results/phylogenetics/all_alignment.fa\\x94\\x8c'results/phylogenetics/all_alignment.csv\\x94\\x8c&results/phylogenetics/deleted_seqs.csv\\x94\\x8c2results/genome_to_comparator/site_identity_map.csv\\x94e}\\x94(\\x8c\\x06_names\\x94}\\x94(\\x8c\\x05trees\\x94K\\x00K\\x03\\x86\\x94\\x8c\\x06states\\x94K\\x03K\\x06\\x86\\x94\\x8c\\talignment\\x94K\\x06N\\x86\\x94\\x8c\\x07all_csv\\x94K\\x07N\\x86\\x94\\x8c\\x0bdeleted_csv\\x94K\\x08N\\x86\\x94\\x8c\\x0ecomparator_map\\x94K\\tN\\x86\\x94u\\x8c\\x12_allowed_overrides\\x94]\\x94(\\x8c\\x05index\\x94\\x8c\\x04sort\\x94eh%\\x8c\\tfunctools\\x94\\x8c\\x07partial\\x94\\x93\\x94h\\x06\\x8c\\x19Namedlist._used_attribute\\x94\\x93\\x94\\x85\\x94R\\x94(h+)}\\x94\\x8c\\x05_name\\x94h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bh\\x17h\\x06\\x8c\\tNamedlist\\x94\\x93\\x94)\\x81\\x94(h\\nh\\x0bh\\x0ce}\\x94(h\\x15}\\x94h#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bubh\\x19h6)\\x81\\x94(h\\rh\\x0eh\\x0fe}\\x94(h\\x15}\\x94h#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bubh\\x1bh\\x10h\\x1dh\\x11h\\x1fh\\x12h!h\\x13ub\\x8c\\x06output\\x94h\\x06\\x8c\\x0bOutputFiles\\x94\\x93\\x94)\\x81\\x94\\x8c\\x0btree_images\\x94a}\\x94(h\\x15}\\x94h#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bub\\x8c\\x06params\\x94h\\x06\\x8c\\x06Params\\x94\\x93\\x94)\\x81\\x94(M\\t\\x01]\\x94(\\x8c\\x14hCoV-19/USA/WA1/2020\\x94\\x8c!hCoV-19/Guangdong/HKU-SZ-002/2020\\x94\\x8c\\x1dhCoV-19/Shandong/LY005-2/2020\\x94e]\\x94(\\x8c\\x06RaTG13\\x94\\x8c\\x06RpYN06\\x94\\x8c\\x06RmYN02\\x94e}\\x94(\\x8c\\x05start\\x94MBT\\x8c\\x03end\\x94Mnsu}\\x94(\\x8c\\x05Wuhan\\x94\\x8c\\x07#56B4E9\\x94\\x8c\\x0bother China\\x94\\x8c\\x07#009E73\\x94\\x8c\\routside China\\x94\\x8c\\x07#F0E442\\x94u}\\x94(\\x8c\\x15Huanan Seafood Market\\x94\\x8c\\x07#0072B2\\x94\\x8c\\x13deleted early Wuhan\\x94\\x8c\\x07#E69F00\\x94\\x8c0Guangdong patient infected in Wuhan before Jan 5\\x94\\x8c\\x07#D55E00\\x94\\x8c\\x0bother Wuhan\\x94\\x8c\\x07#56B4E9\\x94\\x8c\\x0bother China\\x94\\x8c\\x07#009E73\\x94\\x8c\\routside China\\x94\\x8c\\x07#F0E442\\x94ue}\\x94(h\\x15}\\x94(\\x8c\\x0bsite_offset\\x94K\\x00N\\x86\\x94\\x8c\\x0bprogenitors\\x94K\\x01N\\x86\\x94\\x8c\\toutgroups\\x94K\\x02N\\x86\\x94\\x8c\\x12region_of_interest\\x94K\\x03N\\x86\\x94\\x8c\\ncat_colors\\x94K\\x04N\\x86\\x94\\x8c\\rsubcat_colors\\x94K\\x05N\\x86\\x94uh#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bh\\x84M\\t\\x01h\\x86hch\\x88hgh\\x8ahkh\\x8chnh\\x8ehuub\\x8c\\twildcards\\x94h\\x06\\x8c\\tWildcards\\x94\\x93\\x94)\\x81\\x94}\\x94(h\\x15}\\x94h#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bub\\x8c\\x07threads\\x94K\\x01\\x8c\\tresources\\x94h\\x06\\x8c\\tResources\\x94\\x93\\x94)\\x81\\x94(K\\x01K\\x01e}\\x94(h\\x15}\\x94(\\x8c\\x06_cores\\x94K\\x00N\\x86\\x94\\x8c\\x06_nodes\\x94K\\x01N\\x86\\x94uh#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bh\\xafK\\x01h\\xb1K\\x01ub\\x8c\\x03log\\x94h\\x06\\x8c\\x03Log\\x94\\x93\\x94)\\x81\\x94\\x8c,results/logs/notebooks/visualize_trees.ipynb\\x94a}\\x94(h\\x15}\\x94\\x8c\\x08notebook\\x94K\\x00N\\x86\\x94sh#]\\x94(h%h&eh%h)h+\\x85\\x94R\\x94(h+)}\\x94h/h%sNt\\x94bh&h)h+\\x85\\x94R\\x94(h+)}\\x94h/h&sNt\\x94bh\\xc3h\\xc0ub\\x8c\\x06config\\x94}\\x94(\\x8c\\x08max_cpus\\x94K\\x04\\x8c\\x0bscratch_dir\\x94\\x8c\\x11results/_scratch/\\x94\\x8c\\nref_genome\\x94}\\x94(\\x8c\\x05fasta\\x94\\x8c}ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.fna.gz\\x94\\x8c\\x03gff\\x94\\x8c}ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.gff.gz\\x94\\x8c\\x04name\\x94\\x8c\\x07proCoV2\\x94\\x8c\\radd_mutations\\x94]\\x94(\\x8c\\x06C8782T\\x94\\x8c\\x07C18060T\\x94\\x8c\\x07T28144C\\x94eu\\x8c\\x12comparator_genomes\\x94}\\x94(hh}\\x94\\x8c\\x06gisaid\\x94\\x8c\\x0eEPI_ISL_402131\\x94shi}\\x94\\x8c\\x06gisaid\\x94\\x8c\\x0fEPI_ISL_1699446\\x94shj}\\x94\\x8c\\x06gisaid\\x94\\x8c\\x0eEPI_ISL_412977\\x94su\\x8c\\x16gisaid_comparator_dirs\\x94]\\x94\\x8c\\x1edata/comparator_genomes_gisaid\\x94a\\x8c\\nearly_seqs\\x94}\\x94\\x8c\\x06gisaid\\x94}\\x94\\x8c\\x16gisaid_through_Feb2020\\x94\\x8c&data/gisaid_sequences_through_Feb2020/\\x94ss\\x8c\\x16early_seq_header_props\\x94]\\x94(\\x8c\\x06strain\\x94\\x8c\\x0egisaid_epi_isl\\x94\\x8c\\x04date\\x94\\x8c\\x11genbank_accession\\x94\\x8c\\x07country\\x94\\x8c\\x08location\\x94\\x8c\\x03age\\x94\\x8c\\x03sex\\x94e\\x8c\\x08aligners\\x94]\\x94\\x8c\\x08minimap2\\x94a\\x8c\\x04minq\\x94K\\x14\\x8c\\x0fmin_read_length\\x94K\\x14\\x8c\\x16consensus_min_coverage\\x94K\\x03\\x8c\\x12consensus_min_frac\\x94G?\\xe9\\x99\\x99\\x99\\x99\\x99\\x9a\\x8c\\x12region_of_interest\\x94hk\\x8c\\x16who_china_report_cases\\x94\\x8c(data/WHO_China_Report_Dec2019_cases.yaml\\x94\\x8c\\x1awho_china_report_last_date\\x94\\x8c\\x08datetime\\x94\\x8c\\x04date\\x94\\x93\\x94C\\x04\\x07\\xe4\\x01\\x02\\x94\\x85\\x94R\\x94\\x8c\\rwuhan_exports\\x94\\x8c\\x17data/Wuhan_exports.yaml\\x94\\x8c\\x15early_seqs_to_exclude\\x94\\x8c\\x19data/seqs_to_exclude.yaml\\x94\\x8c\\x1dearly_seqs_ignore_muts_before\\x94M\\n\\x01\\x8c\\x1cearly_seqs_ignore_muts_after\\x94M\\xeas\\x8c\\x16early_seqs_filter_runs\\x94}\\x94(\\x8c\\x06n_muts\\x94K\\x04\\x8c\\x04span\\x94K\\nu\\x8c\\x17early_seqs_min_coverage\\x94G?\\xec\\xcc\\xcc\\xcc\\xcc\\xcc\\xcd\\x8c\\x13early_seqs_max_subs\\x94K\\x0f\\x8c\\x18early_seqs_max_ambiguous\\x94M\\x88\\x13\\x8c\\x13early_seqs_max_date\\x94j\\x0b\\x01\\x00\\x00C\\x04\\x07\\xe4\\x02\\x1c\\x94\\x85\\x94R\\x94\\x8c\\x11min_frac_coverage\\x94G?\\xeeffffff\\x8c\\x0fphylo_last_date\\x94j\\x0b\\x01\\x00\\x00C\\x04\\x07\\xe4\\x01\\x1f\\x94\\x85\\x94R\\x94\\x8c\\x15phylo_min_frac_called\\x94G?\\xeeffffff\\x8c\\x14phylo_muts_to_ignore\\x94]\\x94(\\x8c\\x06G3564T\\x94\\x8c\\x06G8790T\\x94\\x8c\\x07G24933T\\x94\\x8c\\x06G2198A\\x94\\x8c\\x06G3145T\\x94\\x8c\\x06A3778G\\x94\\x8c\\x06C6255T\\x94\\x8c\\x06A4050C\\x94\\x8c\\x06T8022G\\x94\\x8c\\x07T13402G\\x94\\x8c\\x07A13947T\\x94\\x8c\\x07A24389C\\x94\\x8c\\x07G24390C\\x94\\x8c\\x06G1149T\\x94\\x8c\\x07C22802G\\x94\\x8c\\x05T153G\\x94e\\x8c\\x18phylo_collapse_rare_muts\\x94K\\x00\\x8c\\x1aphylo_filter_rare_variants\\x94K\\x01\\x8c\\ncat_colors\\x94hn\\x8c\\rsubcat_colors\\x94hu\\x8c\\x14samples_fasterq_dump\\x94}\\x94\\x8c\\x0csamples_wget\\x94}\\x94(\\x8c\\x02A1\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313282\\x94\\x8c\\x0bSRR11313334\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02A2\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313271\\x94\\x8c\\x0bSRR11313461\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02A4\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313417\\x94\\x8c\\x0bSRR11313428\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02A6\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94\\x8c\\x0bSRR11313479\\x94a\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02A9\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313345\\x94\\x8c\\x0bSRR11313356\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03A11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313455\\x94\\x8c\\x0bSRR11313456\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03A12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313473\\x94\\x8c\\x0bSRR11313474\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02B1\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313453\\x94\\x8c\\x0bSRR11313454\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02B2\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313451\\x94\\x8c\\x0bSRR11313452\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02B3\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313448\\x94\\x8c\\x0bSRR11313449\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02B4\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313322\\x94\\x8c\\x0bSRR11313333\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02B9\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313445\\x94\\x8c\\x0bSRR11313446\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02C1\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313287\\x94\\x8c\\x0bSRR11313289\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02C2\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313285\\x94\\x8c\\x0bSRR11313286\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02C9\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313441\\x94\\x8c\\x0bSRR11313442\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03C11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313300\\x94\\x8c\\x0bSRR11313311\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03C12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313443\\x94\\x8c\\x0bSRR11313444\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02D9\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313436\\x94\\x8c\\x0bSRR11313437\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03D11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313438\\x94\\x8c\\x0bSRR11313440\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03D12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313280\\x94\\x8c\\x0bSRR11313281\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02E1\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313276\\x94\\x8c\\x0bSRR11313277\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02E3\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313430\\x94\\x8c\\x0bSRR11313431\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02E5\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313274\\x94\\x8c\\x0bSRR11313275\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02E6\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313427\\x94\\x8c\\x0bSRR11313429\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03E11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313434\\x94\\x8c\\x0bSRR11313435\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03E12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313432\\x94\\x8c\\x0bSRR11313433\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02F5\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313270\\x94\\x8c\\x0bSRR11313470\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02F8\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313423\\x94\\x8c\\x0bSRR11313424\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03F11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313425\\x94\\x8c\\x0bSRR11313426\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02G1\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313468\\x94\\x8c\\x0bSRR11313469\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03G11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313421\\x94\\x8c\\x0bSRR11313422\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03G12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313419\\x94\\x8c\\x0bSRR11313420\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x02H9\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313464\\x94\\x8c\\x0bSRR11313465\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03H12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313466\\x94\\x8c\\x0bSRR11313467\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x10early outpatient\\x94u\\x8c\\x03R01\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313409\\x94\\x8c\\x0bSRR11313410\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R02\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313406\\x94\\x8c\\x0bSRR11313408\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R03\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313269\\x94\\x8c\\x0bSRR11313471\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R04\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94\\x8c\\x0bSRR11313509\\x94a\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R05\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313507\\x94\\x8c\\x0bSRR11313508\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R06\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313505\\x94\\x8c\\x0bSRR11313506\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R07\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313503\\x94\\x8c\\x0bSRR11313504\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R08\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313500\\x94\\x8c\\x0bSRR11313502\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R09\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94\\x8c\\x0bSRR11313498\\x94a\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R10\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313496\\x94\\x8c\\x0bSRR11313497\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R11\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313494\\x94\\x8c\\x0bSRR11313495\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R12\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313492\\x94\\x8c\\x0bSRR11313493\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R13\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313489\\x94\\x8c\\x0bSRR11313491\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R14\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313487\\x94\\x8c\\x0bSRR11313488\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R15\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313485\\x94\\x8c\\x0bSRR11313486\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94u\\x8c\\x03R16\\x94}\\x94(\\x8c\\naccessions\\x94]\\x94(\\x8c\\x0bSRR11313483\\x94\\x8c\\x0bSRR11313484\\x94e\\x8c\\x0ccollected_by\\x94\\x8c#Renmin Hospital of Wuhan University\\x94\\x8c\\rpatient_group\\x94\\x8c\\x16hospital patient (Feb)\\x94uuu\\x8c\\x04rule\\x94\\x8c\\x0fvisualize_trees\\x94\\x8c\\x0fbench_iteration\\x94N\\x8c\\tscriptdir\\x94\\x8c-/home/jbloom/SARS-CoV-2_PRJNA612766/notebooks\\x94ub.\"); from snakemake.logging import logger; logger.printshellcmds = False; import os; os.chdir('/home/jbloom/SARS-CoV-2_PRJNA612766');\n",
    "######## snakemake preamble end #########\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9949903a-a253-442c-a4f5-853281ae8712",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:06:41.342669Z",
     "iopub.status.busy": "2021-06-11T21:06:41.342497Z",
     "iopub.status.idle": "2021-06-11T21:06:41.346634Z",
     "shell.execute_reply": "2021-06-11T21:06:41.345853Z",
     "shell.execute_reply.started": "2021-06-11T21:06:41.342649Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "tree_files = snakemake.input.trees\n",
    "state_files = snakemake.input.states\n",
    "all_csv = snakemake.input.all_csv\n",
    "alignment = snakemake.input.alignment\n",
    "deleted_csv = snakemake.input.deleted_csv\n",
    "comparator_map_csv = snakemake.input.comparator_map\n",
    "site_offset = snakemake.params.site_offset\n",
    "progenitors = snakemake.params.progenitors\n",
    "outgroups = snakemake.params.outgroups\n",
    "region_of_interest = snakemake.params.region_of_interest\n",
    "cat_colors = snakemake.params.cat_colors\n",
    "subcat_colors = snakemake.params.subcat_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24dbcbbf-d480-44d5-88ae-219d22a1d06d",
   "metadata": {},
   "source": [
    "Read data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e378dfd5-7bf6-4976-858e-d761a28e9ddd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:06:41.347693Z",
     "iopub.status.busy": "2021-06-11T21:06:41.347429Z",
     "iopub.status.idle": "2021-06-11T21:06:41.373976Z",
     "shell.execute_reply": "2021-06-11T21:06:41.373410Z",
     "shell.execute_reply.started": "2021-06-11T21:06:41.347673Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_df = pd.read_csv(all_csv, na_filter=None)\n",
    "\n",
    "deleted_df = pd.read_csv(deleted_csv, na_filter=None)\n",
    "\n",
    "outgroup_map = pd.read_csv(comparator_map_csv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409e1979-accd-4de2-816d-92c30aa4851e",
   "metadata": {},
   "source": [
    "Read the reconstructed states and add the tip states, considering only sites where there are multiple identities.\n",
    "Now we make a matrix of substitutions of each node relative to each other.\n",
    "Note that in the general case this could get **really slow**, although for the small-ish alignment of similar sequences used here it's fine:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "76776027-a9f5-4b58-94cc-a6810733eea4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:06:41.374966Z",
     "iopub.status.busy": "2021-06-11T21:06:41.374809Z",
     "iopub.status.idle": "2021-06-11T21:07:21.712937Z",
     "shell.execute_reply": "2021-06-11T21:07:21.711981Z",
     "shell.execute_reply.started": "2021-06-11T21:06:41.374948Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "assert len(state_files) == len(tree_files) == len(progenitors)\n",
    "\n",
    "tip_to_seq = {s.id: list(str(s.seq).upper())\n",
    "              for s in Bio.SeqIO.parse(alignment, 'fasta')}\n",
    "\n",
    "tip_states = (\n",
    "    pd.DataFrame.from_dict(tip_to_seq, orient='index')\n",
    "    .rename_axis('Node')\n",
    "    .reset_index()\n",
    "    .melt(id_vars='Node',\n",
    "          var_name='Site',\n",
    "          value_name='State',\n",
    "          )\n",
    "    .assign(Site=lambda x: x['Site'] + 1)\n",
    "    )\n",
    "\n",
    "subs_matrices = {}\n",
    "\n",
    "for progenitor, state_file in zip(progenitors, state_files):\n",
    "    internal_states = (\n",
    "        pd.read_csv(state_file,\n",
    "                    sep='\\t',\n",
    "                    comment='#',\n",
    "                    usecols=['Node', 'Site', 'State'])\n",
    "        )\n",
    "\n",
    "    states = (\n",
    "        internal_states\n",
    "        .append(tip_states)\n",
    "        .assign(Site=lambda x: x['Site'] + site_offset,\n",
    "                n_states_at_site=lambda x: x.groupby('Site')['State'].transform('nunique'),\n",
    "                )\n",
    "        .query('n_states_at_site > 1')\n",
    "        .drop(columns='n_states_at_site')\n",
    "        )\n",
    "\n",
    "    states_dict = states.set_index(['Node', 'Site'])['State'].to_dict()\n",
    "\n",
    "    nodes = states['Node'].unique().tolist()\n",
    "    sites = sorted(states['Site'].unique())\n",
    "\n",
    "    subs_matrix = {}  # keyed by (parent, descendant)\n",
    "    for n1, n2 in itertools.product(nodes, nodes):\n",
    "        subs = []\n",
    "        for site in sites:\n",
    "            nt1 = states_dict[(n1, site)]\n",
    "            nt2 = states_dict[(n2, site)]\n",
    "            if nt1 != nt2:\n",
    "                if nt1 in {'A', 'C', 'G', 'T'} and nt2 in {'A', 'C', 'G', 'T'}:\n",
    "                    subs.append(f\"{nt1}{site}{nt2}\")\n",
    "        subs_matrix[(n1, n2)] = ', '.join(subs)\n",
    "    subs_matrices[progenitor] = subs_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3871a563-1c6d-4cab-b790-997618387059",
   "metadata": {},
   "source": [
    "Get annotations of which mutations are to an outgroup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "580191ca-cd7e-4ef9-8ffe-86c9904c0dee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:07:21.714330Z",
     "iopub.status.busy": "2021-06-11T21:07:21.714171Z",
     "iopub.status.idle": "2021-06-11T21:07:21.779755Z",
     "shell.execute_reply": "2021-06-11T21:07:21.779127Z",
     "shell.execute_reply.started": "2021-06-11T21:07:21.714312Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "subs_to_outgroup = {outgroup: set(outgroup_map\n",
    "                                  .assign(sub=lambda x: x['site'].astype(str) + x[outgroup])\n",
    "                                  ['sub']\n",
    "                                  )\n",
    "                    for outgroup in outgroups}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de13b90e-f260-417d-82ac-0947e0ecef93",
   "metadata": {},
   "source": [
    "List information on the possible progenitors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a360f121-f971-4595-9f39-145750d91c2f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:07:21.781287Z",
     "iopub.status.busy": "2021-06-11T21:07:21.781146Z",
     "iopub.status.idle": "2021-06-11T21:07:21.816208Z",
     "shell.execute_reply": "2021-06-11T21:07:21.815626Z",
     "shell.execute_reply.started": "2021-06-11T21:07:21.781272Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>date</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substitutions</th>\n",
       "      <th>representative_strain</th>\n",
       "      <th>nstrains</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\"></th>\n",
       "      <th rowspan=\"6\" valign=\"top\">hCoV-19/USA/WA1/2020</th>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/USA/WA1/2020</td>\n",
       "      <td>2020-01-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Guangdong/20SF123/2020</td>\n",
       "      <td>2020-01-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Fujian/8/2020</td>\n",
       "      <td>2020-01-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Guangdong/20SF118/2020</td>\n",
       "      <td>2020-01-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Hangzhou/ZJU-08/2020</td>\n",
       "      <td>2020-01-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Wuhan/0126-C13/2020</td>\n",
       "      <td>2020-01-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"9\" valign=\"top\">T18060C,C29095T</th>\n",
       "      <th rowspan=\"9\" valign=\"top\">hCoV-19/Guangdong/HKU-SZ-002/2020</th>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Guangdong/HKU-SZ-002/2020</td>\n",
       "      <td>2020-01-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Shenzhen/SZTH-002/2020</td>\n",
       "      <td>2020-01-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Guangdong/20SF012/2020</td>\n",
       "      <td>2020-01-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Guangdong/20SF025/2020</td>\n",
       "      <td>2020-01-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Guangdong/20SF013/2020</td>\n",
       "      <td>2020-01-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Shanghai/SH0005/2020</td>\n",
       "      <td>2020-01-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Vietnam/PIHCM-14/2020</td>\n",
       "      <td>2020-01-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Shaoxing/06/2020</td>\n",
       "      <td>2020-01-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hCoV-19/Guangdong/2020XN4276-P0037/2020</td>\n",
       "      <td>2020-01-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">T3171C,T18060C</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">hCoV-19/Shandong/LY005-2/2020</th>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/LY005-2/2020</td>\n",
       "      <td>2020-01-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/2020C1240402/2020</td>\n",
       "      <td>2020-01-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/LY007-2/2020</td>\n",
       "      <td>2020-01-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/LY006-2/2020</td>\n",
       "      <td>2020-01-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/LY004-2/2020</td>\n",
       "      <td>2020-01-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hCoV-19/Shandong/LY008-2/2020</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                             strain  \\\n",
       "substitutions   representative_strain             nstrains                                            \n",
       "                hCoV-19/USA/WA1/2020              6                            hCoV-19/USA/WA1/2020   \n",
       "                                                  6                  hCoV-19/Guangdong/20SF123/2020   \n",
       "                                                  6                           hCoV-19/Fujian/8/2020   \n",
       "                                                  6                  hCoV-19/Guangdong/20SF118/2020   \n",
       "                                                  6                    hCoV-19/Hangzhou/ZJU-08/2020   \n",
       "                                                  6                     hCoV-19/Wuhan/0126-C13/2020   \n",
       "T18060C,C29095T hCoV-19/Guangdong/HKU-SZ-002/2020 9               hCoV-19/Guangdong/HKU-SZ-002/2020   \n",
       "                                                  9                  hCoV-19/Shenzhen/SZTH-002/2020   \n",
       "                                                  9                  hCoV-19/Guangdong/20SF012/2020   \n",
       "                                                  9                  hCoV-19/Guangdong/20SF025/2020   \n",
       "                                                  9                  hCoV-19/Guangdong/20SF013/2020   \n",
       "                                                  9                    hCoV-19/Shanghai/SH0005/2020   \n",
       "                                                  9                   hCoV-19/Vietnam/PIHCM-14/2020   \n",
       "                                                  9                        hCoV-19/Shaoxing/06/2020   \n",
       "                                                  9         hCoV-19/Guangdong/2020XN4276-P0037/2020   \n",
       "T3171C,T18060C  hCoV-19/Shandong/LY005-2/2020     6                   hCoV-19/Shandong/LY005-2/2020   \n",
       "                                                  6              hCoV-19/Shandong/2020C1240402/2020   \n",
       "                                                  6                   hCoV-19/Shandong/LY007-2/2020   \n",
       "                                                  6                   hCoV-19/Shandong/LY006-2/2020   \n",
       "                                                  6                   hCoV-19/Shandong/LY004-2/2020   \n",
       "                                                  6                   hCoV-19/Shandong/LY008-2/2020   \n",
       "\n",
       "                                                                  date  \n",
       "substitutions   representative_strain             nstrains              \n",
       "                hCoV-19/USA/WA1/2020              6         2020-01-19  \n",
       "                                                  6         2020-01-20  \n",
       "                                                  6         2020-01-21  \n",
       "                                                  6         2020-01-21  \n",
       "                                                  6         2020-01-26  \n",
       "                                                  6         2020-01-26  \n",
       "T18060C,C29095T hCoV-19/Guangdong/HKU-SZ-002/2020 9         2020-01-10  \n",
       "                                                  9         2020-01-13  \n",
       "                                                  9         2020-01-14  \n",
       "                                                  9         2020-01-15  \n",
       "                                                  9         2020-01-15  \n",
       "                                                  9         2020-01-28  \n",
       "                                                  9         2020-01-31  \n",
       "                                                  9         2020-01-31  \n",
       "                                                  9         2020-01-31  \n",
       "T3171C,T18060C  hCoV-19/Shandong/LY005-2/2020     6         2020-01-24  \n",
       "                                                  6         2020-01-24  \n",
       "                                                  6         2020-01-25  \n",
       "                                                  6         2020-01-25  \n",
       "                                                  6         2020-01-26  \n",
       "                                                  6         2020-01-30  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', 1000)\n",
    "\n",
    "progenitor_info = (\n",
    "    all_df\n",
    "    .query('representative_strain in @progenitors')\n",
    "    .assign(all_strains_dates=lambda x: x['all_strains_dates'].str.split(', '))\n",
    "    [['substitutions', 'representative_strain', 'nstrains', 'all_strains_dates']]\n",
    "    .set_index(['substitutions', 'representative_strain', 'nstrains'])\n",
    "    .explode('all_strains_dates')\n",
    "    .assign(strain=lambda x: x['all_strains_dates'].str.split(' \\(').str[0],\n",
    "            date=lambda x: x['all_strains_dates'].str.split(' \\(').str[1].str[: -1],\n",
    "            )\n",
    "    .drop(columns='all_strains_dates')\n",
    "    )\n",
    "\n",
    "display(progenitor_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea359c0-bb36-4a53-b786-abf1b584a268",
   "metadata": {},
   "source": [
    "Set up labels on possible progenitors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e6f1fb14-2d3d-4371-85eb-8e9e73aa16d1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:07:21.817366Z",
     "iopub.status.busy": "2021-06-11T21:07:21.817208Z",
     "iopub.status.idle": "2021-06-11T21:07:21.821274Z",
     "shell.execute_reply": "2021-06-11T21:07:21.820795Z",
     "shell.execute_reply.started": "2021-06-11T21:07:21.817348Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "progenitor_dates = progenitor_info.set_index('strain')['date'].to_dict()\n",
    "node_labels = {}\n",
    "for progenitor in progenitors:\n",
    "    node_labels[progenitor] = [progenitor.replace('hCoV-19/', '') +\n",
    "                               f\" ({progenitor_dates[progenitor]})\",\n",
    "                               ]\n",
    "    if progenitor == 'hCoV-19/USA/WA1/2020':\n",
    "        node_labels[progenitor][0] = (node_labels[progenitor][0][: -1] +\n",
    "                                      ', Kumar et al proCoV2)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06cfb783-77cb-4650-a10f-c8ccdff5b723",
   "metadata": {},
   "source": [
    "This next line enables Jupyter notebook rendering of trees in `ete3`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "65e53af0-0e25-460a-a988-6d8e3b55d58e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:07:21.822163Z",
     "iopub.status.busy": "2021-06-11T21:07:21.822003Z",
     "iopub.status.idle": "2021-06-11T21:07:21.824349Z",
     "shell.execute_reply": "2021-06-11T21:07:21.823885Z",
     "shell.execute_reply.started": "2021-06-11T21:07:21.822144Z"
    }
   },
   "outputs": [],
   "source": [
    "# to enable jupyter notebook rendering: \n",
    "os.environ['QT_QPA_PLATFORM'] = 'offscreen'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b87da384-5573-487b-9082-99c52ce11eaa",
   "metadata": {},
   "source": [
    "Add in deleted sequences at all \"plausible\" locations, adding proportional to number of nodes at each location at which it's consistent, and making the category labels the sub categories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "521552ac-cca7-4c5a-9753-8cc4528560cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:07:21.825220Z",
     "iopub.status.busy": "2021-06-11T21:07:21.825065Z",
     "iopub.status.idle": "2021-06-11T21:07:21.861174Z",
     "shell.execute_reply": "2021-06-11T21:07:21.860726Z",
     "shell.execute_reply.started": "2021-06-11T21:07:21.825202Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 11 sequences comprising 79 strains compatible with the 3 deleted sequences with the following substitutions: \n",
      "There are 1 sequences comprising 2 strains compatible with the 1 deleted sequences with the following substitutions: C24034T,T26729C,G28077C\n",
      "There are 0 sequences comprising 0 strains compatible with the 1 deleted sequences with the following substitutions: C25460T,C28144T\n",
      "There are 16 sequences comprising 113 strains compatible with the 4 deleted sequences with the following substitutions: C28144T\n",
      "There are 0 sequences comprising 0 strains compatible with the 1 deleted sequences with the following substitutions: C28144T,G28514T\n",
      "There are 3 sequences comprising 14 strains compatible with the 1 deleted sequences with the following substitutions: C29095T\n",
      "There are 0 sequences comprising 0 strains compatible with the 1 deleted sequences with the following substitutions: G22081A,C28144T,T29483G\n",
      "There are 0 sequences comprising 0 strains compatible with the 1 deleted sequences with the following substitutions: T25304A,C28144T\n"
     ]
    }
   ],
   "source": [
    "start, end = region_of_interest['start'], region_of_interest['end']\n",
    "\n",
    "def is_compatible(all_subs, subs):\n",
    "    all_subs = {s for s in all_subs.split(',')\n",
    "                if s and start <= int(s[1 : -1]) <= end}\n",
    "    return subs == all_subs\n",
    "\n",
    "all_plus_deleted_df = (\n",
    "    all_df\n",
    "    .assign(deleted=0)\n",
    "    .rename(columns={c: c.split('_')[1] if c.startswith('subcat_') else c\n",
    "                     for c in all_df.columns})\n",
    "    )\n",
    "for tup in deleted_df.itertuples():\n",
    "    subs_str, n = tup.substitutions, tup.nstrains\n",
    "    subs = {s for s in subs_str.split(',') if s}\n",
    "    all_plus_deleted_df = (\n",
    "        all_plus_deleted_df\n",
    "        .assign(compatible=lambda x: x['substitutions'].apply(is_compatible, args=(subs,)))\n",
    "        )\n",
    "    compatible_df = all_plus_deleted_df[['substitutions', 'nstrains', 'compatible']].query('compatible')\n",
    "    n_tot = compatible_df['nstrains'].sum()\n",
    "    print(f\"There are {len(compatible_df)} sequences comprising {n_tot} strains compatible \"\n",
    "          f\"with the {n} deleted sequences with the following substitutions: {subs_str}\")\n",
    "    if n_tot == 0:\n",
    "        continue\n",
    "    all_plus_deleted_df = (\n",
    "        all_plus_deleted_df\n",
    "        .assign(deleted=lambda x: x['deleted'] + x['nstrains'] * n / n_tot * x['compatible'].astype(int))\n",
    "        )\n",
    "all_plus_deleted_df = (\n",
    "    all_plus_deleted_df\n",
    "    .assign(nstrains=lambda x: x['nstrains'] + x['deleted'])\n",
    "    .rename(columns={'deleted': 'deleted early Wuhan'})\n",
    "    .drop(columns='compatible')\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d836f4df-3014-4eb6-9477-c5b7ae2b977d",
   "metadata": {},
   "source": [
    "Modify `all_df` to just have categories as columns, and also to split the `Wuhan` category into `Huanan Seafood Market` and `other Wuhan`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c1bed3f6-eb3c-4509-9855-8f3d04782f28",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:35:57.639992Z",
     "iopub.status.busy": "2021-06-11T21:35:57.639564Z",
     "iopub.status.idle": "2021-06-11T21:35:57.663056Z",
     "shell.execute_reply": "2021-06-11T21:35:57.662154Z",
     "shell.execute_reply.started": "2021-06-11T21:35:57.639946Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "convert_cats_to_subcats = {'Wuhan': ['Huanan Seafood Market', 'other Wuhan']}\n",
    "\n",
    "for cat, subcats in convert_cats_to_subcats.items():\n",
    "    assert all(all_df[f\"cat_{cat}\"] == all_df[[f\"subcat_{subcat}\" for subcat in subcats]].sum(axis=1))\n",
    "    all_df_cat_cols = (\n",
    "        all_df\n",
    "        .drop(columns=f\"cat_{cat}\")\n",
    "        .rename(columns={f\"subcat_{subcat}\": subcat.replace('subcat_', 'cat_')\n",
    "                         for subcat in subcats}\n",
    "                )\n",
    "        )\n",
    "    new_cat_colors = {}\n",
    "    for key, val in cat_colors.items():\n",
    "        if key == cat:\n",
    "            for subcat in subcats:\n",
    "                new_cat_colors[subcat] = subcat_colors[subcat]\n",
    "        else:\n",
    "            new_cat_colors[key] = val\n",
    "    cat_colors = new_cat_colors\n",
    "\n",
    "all_df_cat_cols = all_df.rename(columns={c: c.split('_')[1] if c.startswith('cat_') else c\n",
    "                                         for c in all_df.columns})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a53105-65a7-46fd-9d21-bfaeee76f328",
   "metadata": {},
   "source": [
    "Draw the trees:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "471a0505-cd4c-4762-88a4-601ea5a102bd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-11T21:36:12.811155Z",
     "iopub.status.busy": "2021-06-11T21:36:12.810610Z",
     "iopub.status.idle": "2021-06-11T21:36:12.861140Z",
     "shell.execute_reply": "2021-06-11T21:36:12.859994Z",
     "shell.execute_reply.started": "2021-06-11T21:36:12.811105Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "annotations missing Huanan Seafood Market:\n{'substitutions': '', 'nstrains': 6, 'all_strains': 'hCoV-19/USA/WA1/2020, hCoV-19/Guangdong/20SF123/2020, hCoV-19/Fujian/8/2020, hCoV-19/Guangdong/20SF118/2020, hCoV-19/Hangzhou/ZJU-08/2020, hCoV-19/Wuhan/0126-C13/2020', 'all_strains_dates': 'hCoV-19/USA/WA1/2020 (2020-01-19), hCoV-19/Guangdong/20SF123/2020 (2020-01-20), hCoV-19/Fujian/8/2020 (2020-01-21), hCoV-19/Guangdong/20SF118/2020 (2020-01-21), hCoV-19/Hangzhou/ZJU-08/2020 (2020-01-26), hCoV-19/Wuhan/0126-C13/2020 (2020-01-26)', 'Wuhan': 1, 'other China': 4, 'outside China': 1, 'subcat_Huanan Seafood Market': 0, 'subcat_other Wuhan': 1, 'subcat_other China': 4, 'subcat_outside China': 1, 'subcat_Guangdong patient infected in Wuhan before Jan 5': 0, 'RaTG13_delta_dist': 0, 'RpYN06_delta_dist': 0, 'RmYN02_delta_dist': 0}",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-45-6784eb33a356>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    206\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0moutgroup\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'RaTG13'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m         \u001b[0;32mcontinue\u001b[0m  \u001b[0;31m# debugging\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 208\u001b[0;31m     t, ts = get_pretty_tree(tree_file,\n\u001b[0m\u001b[1;32m    209\u001b[0m                             \u001b[0mdf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mall_df_cat_cols\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    210\u001b[0m                             \u001b[0msubs_matrix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msubs_matrices\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprogenitor\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-45-6784eb33a356>\u001b[0m in \u001b[0;36mget_pretty_tree\u001b[0;34m(treefile, df, subs_matrix, progenitor, to_outgroup_muts, locations, widthscale, heightscale, nodesizescale, label_nodes, label_fontsize, branch_linewidth, mut_label_color, mut_to_outgroup_label_color, collapse_below, node_label_color, node_labels, title_fontsize)\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mloc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlocations\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mloc\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mannotations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"annotations missing {loc}:\\n{list(annotations.values())[0]}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0mt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mete3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtreefile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: annotations missing Huanan Seafood Market:\n{'substitutions': '', 'nstrains': 6, 'all_strains': 'hCoV-19/USA/WA1/2020, hCoV-19/Guangdong/20SF123/2020, hCoV-19/Fujian/8/2020, hCoV-19/Guangdong/20SF118/2020, hCoV-19/Hangzhou/ZJU-08/2020, hCoV-19/Wuhan/0126-C13/2020', 'all_strains_dates': 'hCoV-19/USA/WA1/2020 (2020-01-19), hCoV-19/Guangdong/20SF123/2020 (2020-01-20), hCoV-19/Fujian/8/2020 (2020-01-21), hCoV-19/Guangdong/20SF118/2020 (2020-01-21), hCoV-19/Hangzhou/ZJU-08/2020 (2020-01-26), hCoV-19/Wuhan/0126-C13/2020 (2020-01-26)', 'Wuhan': 1, 'other China': 4, 'outside China': 1, 'subcat_Huanan Seafood Market': 0, 'subcat_other Wuhan': 1, 'subcat_other China': 4, 'subcat_outside China': 1, 'subcat_Guangdong patient infected in Wuhan before Jan 5': 0, 'RaTG13_delta_dist': 0, 'RpYN06_delta_dist': 0, 'RmYN02_delta_dist': 0}"
     ]
    }
   ],
   "source": [
    "def get_PieChartFace(node_annotation, nodesizescale, locations):\n",
    "    nstrains = node_annotation['nstrains']\n",
    "    loc_sum = sum(node_annotation[loc] for loc in locations)\n",
    "    assert numpy.allclose(nstrains, loc_sum, atol=1e-3), f\"{nstrains=}\\n{loc_sum=}\\n{locations=}\\n{node_annotation=}\"\n",
    "    radius = nodesizescale * math.sqrt(node_annotation['nstrains'])\n",
    "    percents = []\n",
    "    colors = []\n",
    "    for loc, color in locations.items():\n",
    "        percents.append(100 * node_annotation[loc] / nstrains)\n",
    "        colors.append(color)\n",
    "    return ete3.PieChartFace(percents=percents,\n",
    "                             colors=colors,\n",
    "                             width=radius,\n",
    "                             height=radius,\n",
    "                             #line_color='black',\n",
    "                             )\n",
    "\n",
    "def get_pretty_tree(treefile,\n",
    "                    df,\n",
    "                    subs_matrix,\n",
    "                    progenitor,\n",
    "                    to_outgroup_muts,\n",
    "                    locations,\n",
    "                    widthscale=1350,\n",
    "                    heightscale=1,\n",
    "                    nodesizescale=35,\n",
    "                    label_nodes=False,\n",
    "                    label_fontsize=25,\n",
    "                    branch_linewidth=4,\n",
    "                    mut_label_color='black',\n",
    "                    mut_to_outgroup_label_color='#CC79A7',\n",
    "                    collapse_below={'C28144T': ['colapsed clade B', 'Wuhan-Hu-1']},\n",
    "                    node_label_color='#999999',\n",
    "                    node_labels={},\n",
    "                    title_fontsize=35,\n",
    "                    ):\n",
    "    \"\"\"Returns `(tree, tree_style)`.\"\"\"\n",
    "    annotations = df.set_index('representative_strain').to_dict(orient='index')\n",
    "    for loc in locations:\n",
    "        if loc not in list(annotations.values())[0]:\n",
    "            raise ValueError(f\"annotations missing {loc}:\\n{list(annotations.values())[0]}\")\n",
    "   \n",
    "    t = ete3.Tree(treefile, format=1)\n",
    "\n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = False  # add tip names manually\n",
    "     \n",
    "    nstyle_dict = {'hz_line_width': branch_linewidth,\n",
    "                   'vt_line_width': branch_linewidth,\n",
    "                   'hz_line_color': 'black',\n",
    "                   'vt_line_color': 'black',\n",
    "                   'size': 0}\n",
    "\n",
    "    # label nodes\n",
    "    for n in t.traverse():\n",
    "        if n != t:\n",
    "            subs = subs_matrix[(n.up.name, n.name)]\n",
    "            icol = 0\n",
    "            for sub in subs.split(', '):\n",
    "                if sub:\n",
    "                    to_outgroup = sub[1:] in to_outgroup_muts\n",
    "                    n.add_face(ete3.TextFace(f\"{sub}  \",\n",
    "                                             fsize=label_fontsize,\n",
    "                                             fgcolor=(mut_to_outgroup_label_color\n",
    "                                                      if to_outgroup else\n",
    "                                                      mut_label_color),\n",
    "                                             bold=to_outgroup,\n",
    "                                             ),\n",
    "                               column=icol,\n",
    "                               position='branch-top',\n",
    "                               )\n",
    "                    icol += 1\n",
    "        nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "        if n.is_leaf() or n.name == progenitor:\n",
    "            n.add_face(get_PieChartFace(annotations[n.name], nodesizescale, locations),\n",
    "                       column=0,\n",
    "                       position='branch-right',\n",
    "                       )\n",
    "            if label_nodes:\n",
    "                n.add_face(ete3.TextFace(annotations[n.name]['substitutions'],\n",
    "                                         tight_text=True,\n",
    "                                         fsize=labelfontsize,\n",
    "                                         ),\n",
    "                           column=0,\n",
    "                           position='branch-right',\n",
    "                           )\n",
    "        n.set_style(nstyle)\n",
    "        \n",
    "    # set dummy node for root, this allows us to put\n",
    "    # progenitor at base of root branch\n",
    "    progenitor_n = t.search_nodes(name=progenitor)\n",
    "    assert len(progenitor_n) == 1\n",
    "    progenitor_n = progenitor_n[0]\n",
    "    dummy_outgroup = progenitor_n.add_child(name='dummy',\n",
    "                                            dist=1e-5)\n",
    "    t.set_outgroup(dummy_outgroup)\n",
    "    assert len(progenitor_n.children) == 1\n",
    "    progenitor_child = progenitor_n.children[0]\n",
    "    nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "    progenitor_child.set_style(nstyle)\n",
    "    if progenitor_n.faces:\n",
    "        sub = subs_matrix[(progenitor_n.up.name, progenitor)]\n",
    "        if sub:\n",
    "            assert len(sub.split(',')) == 1\n",
    "            sub_flipped = f\"{sub[-1]}{sub[1: -1]}{sub[0]}\"\n",
    "            to_outgroup = sub_flipped[1:] in to_outgroup_muts\n",
    "            progenitor_child.add_face(ete3.TextFace(f\"{sub_flipped}  \",\n",
    "                                                    fsize=label_fontsize,\n",
    "                                                    fgcolor=(mut_to_outgroup_label_color\n",
    "                                                             if to_outgroup else\n",
    "                                                             mut_label_color),\n",
    "                                                    bold=to_outgroup,\n",
    "                                                    ),\n",
    "                                      column=0,\n",
    "                                      position='branch-top',\n",
    "                                      )\n",
    "        delattr(progenitor_n, '_faces')\n",
    "    empty_nstyle = ete3.NodeStyle(hz_line_color='white',\n",
    "                                  vt_line_color='white',\n",
    "                                  size=0)\n",
    "    dummy_outgroup.set_style(empty_nstyle)\n",
    "    dummy_outgroup.up.set_style(empty_nstyle)\n",
    "    progenitor_n.add_face(get_PieChartFace(annotations[progenitor], nodesizescale, locations),\n",
    "                          column=0,\n",
    "                          position='branch-right',\n",
    "                          )\n",
    "\n",
    "    t.ladderize()\n",
    "    \n",
    "    # collapse tree at indicated nodes\n",
    "    for below_subs, txt in collapse_below.items():\n",
    "        collapse_n = [n for n in t.traverse() if n.name and n.up and n.up.name and\n",
    "                      n.name != 'dummy' and subs_matrix[(n.up.name, n.name)] == below_subs]\n",
    "        if len(collapse_n) > 1:\n",
    "            raise ValueError(f\"more than one node to collapse for {below_subs}\")\n",
    "        elif len(collapse_n) == 0:\n",
    "            raise ValueError(f\"no nodes to collapse for {below_subs}\")\n",
    "        collapse_n = collapse_n[0]\n",
    "        collapse_nstrains = 0\n",
    "        collapse_annotations = {key: 0 for key in ['nstrains', *locations]}\n",
    "        for n in [collapse_n] + collapse_n.get_descendants():\n",
    "            if n.is_leaf():\n",
    "                collapse_annotations['nstrains'] += annotations[n.name]['nstrains']\n",
    "                for location in locations:\n",
    "                    collapse_annotations[location] += annotations[n.name][location]\n",
    "            if n != collapse_n:\n",
    "                n.detach()\n",
    "        nstyle = ete3.NodeStyle(**nstyle_dict)\n",
    "        collapse_n.set_style(nstyle)\n",
    "        collapse_n.add_face(get_PieChartFace(collapse_annotations, nodesizescale, locations),\n",
    "                            column=0,\n",
    "                            position='branch-right',\n",
    "                            )\n",
    "        for face_txt in txt:\n",
    "            collapse_n.add_face(ete3.TextFace(f\" {face_txt}\",\n",
    "                                              fsize=label_fontsize,\n",
    "                                              fgcolor=node_label_color,\n",
    "                                              bold=True,\n",
    "                                              ),\n",
    "                                column=1,\n",
    "                                position='branch-right',\n",
    "                                )\n",
    "    \n",
    "    # label specified non-progenitor nodes\n",
    "    for name, labels in node_labels.items():\n",
    "        if name == progenitor:\n",
    "            continue\n",
    "        n = t.search_nodes(name=name)\n",
    "        assert len(n) == 1\n",
    "        n = n[0]\n",
    "        for face_txt in labels:\n",
    "            n.add_face(ete3.TextFace(f\" {face_txt}\",\n",
    "                                     fsize=label_fontsize,\n",
    "                                     fgcolor=node_label_color,\n",
    "                                     bold=True,\n",
    "                                     ),\n",
    "                       column=1,\n",
    "                       position='branch-right',\n",
    "                       )\n",
    "            \n",
    "    # add title\n",
    "    if progenitor in node_labels:\n",
    "        title_labels = node_labels[progenitor]\n",
    "    else:\n",
    "        title_labels = [progenitor]\n",
    "    for face_txt in title_labels:\n",
    "        ts.title.add_face(ete3.TextFace('progenitor as ' + face_txt,\n",
    "                                        fsize=title_fontsize,\n",
    "                                        fgcolor='black'),\n",
    "                          column=0,\n",
    "                          )\n",
    "    ts.title.add_face(ete3.TextFace('', fsize=title_fontsize),\n",
    "                      column=0)\n",
    "    \n",
    "    ts.show_scale = False\n",
    "    height = t.get_farthest_node()[1]\n",
    "    ts.scale = widthscale / height\n",
    "    ts.branch_vertical_margin = heightscale\n",
    "    \n",
    "    return t, ts\n",
    "\n",
    "assert len(progenitors) == len(tree_files)\n",
    "for outgroup, (progenitor, tree_file) in itertools.product(outgroups,\n",
    "                                                           zip(progenitors, tree_files)\n",
    "                                                           ):\n",
    "    if outgroup != 'RaTG13':\n",
    "        continue  # debugging\n",
    "    t, ts = get_pretty_tree(tree_file,\n",
    "                            df=all_df_cat_cols,\n",
    "                            subs_matrix=subs_matrices[progenitor],\n",
    "                            progenitor=progenitor,\n",
    "                            to_outgroup_muts=subs_to_outgroup[outgroup],\n",
    "                            node_labels=node_labels,\n",
    "                            locations=cat_colors,\n",
    "                            )\n",
    "    display(t.render('%%inline', tree_style=ts, w=300))\n",
    "    t_deleted, ts_deleted = get_pretty_tree(\n",
    "                            tree_file,\n",
    "                            df=all_plus_deleted_df,\n",
    "                            subs_matrix=subs_matrices[progenitor],\n",
    "                            progenitor=progenitor,\n",
    "                            to_outgroup_muts=subs_to_outgroup[outgroup],\n",
    "                            node_labels=node_labels,\n",
    "                            locations=subcat_colors,\n",
    "                            )\n",
    "    display(t_deleted.render('%%inline', tree_style=ts_deleted, w=300))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "040049df-28d0-485e-8481-253cca700016",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
